rest/session_api.go

```diff
@@ -166,45 +168,43 @@ func (h *handler) makeSessionFromEmail(email string, createUserIfNeeded bool) er
 // If not found, and createUserIfNeeded=true, creates a new user based on username, email.
 func (h *handler) makeSessionFromNameAndEmail(username, email string, createUserIfNeeded bool) error {
 
-	// Username and email are verified. Look up the user and make a login session for her - first
-	// attempt lookup by name
+	// First attempt lookup by username and make a login session for her.
 	user, err := h.db.Authenticator().GetUser(username)
 	if err != nil {
 		return err
 	}
 
-	// If user found, check whether the email needs to be updated (e.g. user has changed email in
-	// external auth system)
-	if user != nil {
-		if email != user.Email() {
-			if err := user.SetEmail(email); err == nil {
+	// Attempt email updates/lookups if an email is provided.
+	if len(email) > 0 {
+		// If user found, check whether the email needs to be updated
+		// (e.g. user has changed email in external auth system)
+		if user != nil && email != user.Email() {
+			if err = user.SetEmail(email); err == nil {
 				h.db.Authenticator().Save(user)
 			}
 		}
-	} else {
-		// User not found by name.  Attempt user lookup by email.  This provides backward
+
```

Comment:
I think you lost an 'else' here - we don't want to fall through to the email-based user lookup unless user was nil.